---
title: "Sample_09_03_Tomato"
author: "Irepan Salvador-Martinez"
date: "18/05/2021"
output:
  BiocStyle::html_document:
    toc: true
    toc_float: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
  echo = TRUE, out.width = "100%", fig.align = "center",fig.width = 9,fig.height = 6,
  message = FALSE, warning = FALSE
)
options(width = 1200)
```


# 09_03  Samples ADULT MICE

## About the samples (from Rui's email)

With the 09_03 samples we will finally be able to compare Lineages of control (YFP+) vs Rbpj LOF (Mutant) samples.

In this case, and as said before, we expect Rbpj LOF cells to make all EC lineages well, except the ARTERIAL side of the vascular tree, that should be formed mainly by YFP+ cells. So I expect that in the case of this Tomato+ vs YFP+ competition for vascular development, the Rbpj LOF will form most of the venous side of the organs, and the YFP+ mostly the arterial side of the vascular tree.

From a scRNAseq plot showing a Control vs inducible Dll4/Notch1/Rbpj loss of function in Liver Adult vessels for only 2 weeks it can be seen that the mutant cells, loose the arterial identity and instead become more venous and highly metabolic/proliferative.
This data (not shown) has nothing on DEVELOPMENT/Lineages, is just their role in the HOMEOSTATIC Liver ArterioVenous zonation.

### Some of the MARKER genes.
RSPO3 (Veins), Wnt2 (Venous capillaries), LTBP4 (Arterial capillaries), Gja5 (Arteries), Odc1 (Dll4 or Rbpj LOF Highly Metabolic cells), Mki67 (Proliferating Cells in G2/M) and MCM6 for cells in S-phase.

### About the analysis
After a first analysis of the data, in the case of the 09_03 analysis we really need more cellular/cluster resolution per organ than before.
We need similar cluster resolution to the one for Adult Liver ECs performed by Carlos at CNIC.
Not sure if you can input the same parameters he did in order to do this clustering below.

For the 09_03 samples, we would like to see the YFP+ and Tomato+ ECs plotted together in the same Umap/tSNE space, to see how they segregate developmentally in the Arteriovenous axis as I suggest above.

**NOTE** This notebook is based on the [Seurat tutorial found here](https://satijalab.org/seurat/archive/v3.2/pbmc3k_tutorial.html)

# About input files

The files used as input for this analysis are the files coming out of the the
*cellranger count* pipeline, which essentialy maps the raw reads into the genome
and takes into account the 10x UMIs and cell barcodes to assign the reads to 
the different cells, genes, and molecules.

In this case I used cellranger v6.0.0

The cellranger command is the usual one but in order to read our Recombinase
arrays and other transgenes (YFP, Tomato, etc) I needed to append their sequences into the fasta files and the IDs into
the gtf files (using the *cellranger mkref* pipeline).


```{r packages}
library(dplyr)
library(Seurat)
library(patchwork)
library(ggplot2)
library(DoubletFinder)
library(BiocStyle)
library(dittoSeq)
```

## Read 10X data

We start by reading in the data. The Read10X function reads in the output of the cellranger pipeline from 10X, returning a unique molecular identified (UMI) count matrix. The values in this matrix represent the number of molecules for each feature (i.e. gene; row) that are detected in each cell (column).

We next use the count matrix to create a Seurat object. The object serves as a container that contains both data (like the count matrix) and analysis (like PCA, or clustering results) for a single-cell dataset. For a technical discussion of the Seurat object structure, check out our GitHub Wiki. For example, the count matrix is stored in Sample_09_03_Tomato[["RNA"]]@counts.

```{r Load precomputed data}
Sample_09_03_Tomato <- readRDS("Sample_09_03_Tomato.rds")
Sample_09_03_Tomato_nonblood <- readRDS("Sample_09_03_Tomato_nonblood.rds")
```

Here, we analyze a dataset where transcriptomic measurements are paired with abundance estimates for 1 surface protein, whose levels are quantified with DNA-barcoded antibody. 
First, we load in two count matrices : one for the RNA measurements, and one for the antibody-derived tags (ADT). 

```{r Read 10X data,eval=FALSE}
# Load the dataset
path_to_outs <- "../cellranger_outs/Sample_09_03_Tomato/"
Sample_09_03_Tomato.data <- Read10X(data.dir = paste(path_to_outs, "filtered_feature_bc_matrix/", sep = ""))

# creates a Seurat object based on the scRNA-seq data
Sample_09_03_Tomato <- CreateSeuratObject(counts = Sample_09_03_Tomato.data, 
                                       project = "Sample_09_03_Tomato", 
                                       min.cells = 3 )#, min.features = 200)
rm(Sample_09_03_Tomato.data)
```

# Standard pre-processing workflow

The steps below encompass the standard pre-processing workflow for scRNA-seq data in Seurat. These represent the selection and filtration of cells based on QC metrics, data normalization and scaling, and the detection of highly variable features.

### QC and selecting cells for further analysis

Seurat allows you to easily explore QC metrics and filter cells based on any user-defined criteria. A few QC metrics commonly used by the community include

  The number of unique genes detected in each cell.
    Low-quality cells or empty droplets will often have very few genes
    Cell doublets or multiplets may exhibit an aberrantly high gene count
  Similarly, the total number of molecules detected within a cell (correlates strongly with unique genes)
  The percentage of reads that map to the mitochondrial genome
    Low-quality / dying cells often exhibit extensive mitochondrial contamination
    We calculate mitochondrial QC metrics with the PercentageFeatureSet function, which calculates the percentage of counts originating   from a set of features
    We use the set of all genes starting with MT- as a set of mitochondrial genes

```{r Mt genes,eval=FALSE}
# The [[ operator can add columns to object metadata. This is a great place to stash QC stats
Sample_09_03_Tomato[["percent.mt"]] <- PercentageFeatureSet(Sample_09_03_Tomato, pattern = "^mt-")
# Show QC metrics for the first 5 cells
head(Sample_09_03_Tomato@meta.data, 50)
```


In the example below, we visualize QC metrics, and use these to filter cells.

  We filter cells that have unique feature counts over 2,500 or less than 200
  We filter cells that have >10% mitochondrial counts



```{r QC metrics}
# Visualize QC metrics as a violin plot
VlnPlot(Sample_09_03_Tomato, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"),
        group.by = "orig.ident", ncol = 3, pt.size = 0)
```



```{r plot features}
# FeatureScatter is typically used to visualize feature-feature relationships, but can be used
# for anything calculated by the object, i.e. columns in object metadata, PC scores etc.

plot1 <- FeatureScatter(Sample_09_03_Tomato, feature1 = "nFeature_RNA",
                        group.by = "orig.ident", feature2 = "percent.mt") +
          geom_vline(xintercept = c(200,5000),linetype = 2 ) +
          geom_hline(yintercept = 15 ,linetype = 2)
plot2 <- FeatureScatter(Sample_09_03_Tomato, feature1 = "nCount_RNA",
                        group.by = "orig.ident",feature2 = "nFeature_RNA")+
          geom_hline(yintercept = c(200,5000),linetype = 2 )
plot1 / plot2
```



```{r Filter cells,eval=FALSE}
Sample_09_03_Tomato <- subset(Sample_09_03_Tomato, 
                           subset = nFeature_RNA > 200 & nFeature_RNA < 5000 & percent.mt < 15)
```

## Normalizing the data

After removing unwanted cells from the dataset, the next step is to normalize the data. By default, we employ a global-scaling normalization method "LogNormalize" that normalizes the feature expression measurements for each cell by the total expression, multiplies this by a scale factor (10,000 by default), and log-transforms the result. Normalized values are stored in Sample_09_03_Tomato[["RNA"]]@data.

```{r Normalise, eval=FALSE}
Sample_09_03_Tomato <- NormalizeData(Sample_09_03_Tomato, 
                                  normalization.method = "LogNormalize",
                                  scale.factor = 10000)

```

## Identification of highly variable features (feature selection)

We next calculate a subset of features that exhibit high cell-to-cell variation in the dataset (i.e, they are highly expressed in some cells, and lowly expressed in others). We and others have found that focusing on these genes in downstream analysis helps to highlight biological signal in single-cell datasets.

Our procedure in Seurat3 is described in detail here, and improves on previous versions by directly modeling the mean-variance relationship inherent in single-cell data, and is implemented in the FindVariableFeatures function. By default, we return 2,000 features per dataset. These will be used in downstream analysis, like PCA.


```{r Highly variable genes,eval=FALSE}

Sample_09_03_Tomato <- FindVariableFeatures(Sample_09_03_Tomato, 
                                         selection.method = "vst",
                                         nfeatures = 2000)

```

```{r SHOW Highly variable genes,eval=TRUE}
# Identify the 10 most highly variable genes
top10 <- head(VariableFeatures(Sample_09_03_Tomato), 10)

# plot variable features with and without labels
plot1 <- VariableFeaturePlot(Sample_09_03_Tomato)
plot2 <- LabelPoints(plot = plot1, points = top10, repel = TRUE)
plot2

```

## Scaling the data

Next, we apply a linear transformation ('scaling') that is a standard pre-processing step prior to dimensional reduction techniques like PCA. The ScaleData function:

  Shifts the expression of each gene, so that the mean expression across cells is 0
  Scales the expression of each gene, so that the variance across cells is 1
    This step gives equal weight in downstream analyses, so that highly-expressed genes do not dominate
  The results of this are stored in Sample_09_03_Tomato[["RNA"]]@scale.data
```{r Scaling, eval=FALSE}
all.genes <- rownames(Sample_09_03_Tomato)
Sample_09_03_Tomato <- ScaleData(Sample_09_03_Tomato, features = all.genes)

```

## Perform linear dimensional reduction

Next we perform PCA on the scaled data. By default, only the previously determined variable features are used as input, but can be defined using features argument if you wish to choose a different subset.

```{r PCA, eval=FALSE}
Sample_09_03_Tomato <- RunPCA(Sample_09_03_Tomato,
                           features = VariableFeatures(object = Sample_09_03_Tomato))
```

Seurat provides several useful ways of visualizing both cells and features that define the PCA, including VizDimReduction, DimPlot, and DimHeatmap

```{r}
# Examine and visualize PCA results a few different ways
print(Sample_09_03_Tomato[["pca"]], dims = 1:5, nfeatures = 5)
```

```{r PCA genes}
VizDimLoadings(Sample_09_03_Tomato, dims = 1:2, reduction = "pca")
```


```{r PCA viz}
DimPlot(Sample_09_03_Tomato,group.by = "orig.ident", reduction = "pca")
```

```{r PCA 12 dims}
DimHeatmap(Sample_09_03_Tomato, dims = 1:9, cells = 500, balanced = TRUE)
```

## Determine the 'dimensionality' of the dataset

To overcome the extensive technical noise in any single feature for scRNA-seq data, Seurat clusters cells based on their PCA scores, with each PC essentially representing a 'metafeature' that combines information across a correlated feature set. The top principal components therefore represent a robust compression of the dataset. However, how many componenets should we choose to include? 10? 20? 100?

From Seurat:
"In Macosko et al, we implemented a resampling test inspired by the JackStraw procedure. We randomly permute a subset of the data (1% by default) and rerun PCA, constructing a 'null distribution' of feature scores, and repeat this procedure. We identify 'significant' PCs as those who have a strong enrichment of low p-value features."

```{r PCA dimensionality, eval=FALSE}
# NOTE: This process can take a long time for big datasets, comment out for expediency. More
# approximate techniques such as those implemented in ElbowPlot() can be used to reduce
# computation time
Sample_09_03_Tomato <- JackStraw(Sample_09_03_Tomato, num.replicate = 100, dims = 30)
Sample_09_03_Tomato <- ScoreJackStraw(Sample_09_03_Tomato, dims = 1:30)
```

The JackStrawPlot function provides a visualization tool for comparing the distribution of p-values for each PC with a uniform distribution (dashed line). 'Significant' PCs will show a strong enrichment of features with low p-values (solid curve above the dashed line). In this case it appears that there is a drop-off in significance after the first 20-25 PCs.

```{r PCA dimensionality plot}
JackStrawPlot(Sample_09_03_Tomato, dims = 1:30)

```

```{r Elbowplot} 
ElbowPlot(Sample_09_03_Tomato, ndims = 30)
```

## Cluster the cells

Seurat v3 applies a graph-based clustering approach, building upon initial strategies in (Macosko et al). Importantly, the distance metric which drives the clustering analysis (based on previously identified PCs) remains the same. However, the approach to partioning the cellular distance matrix into clusters has dramatically improved. The approach was heavily inspired by recent manuscripts which applied graph-based clustering approaches to scRNA-seq data [SNN-Cliq, Xu and Su, Bioinformatics, 2015] and CyTOF data [PhenoGraph, Levine et al., Cell, 2015]. Briefly, these methods embed cells in a graph structure - for example a K-nearest neighbor (KNN) graph, with edges drawn between cells with similar feature expression patterns, and then attempt to partition this graph into highly interconnected 'quasi-cliques' or 'communities'.

As in PhenoGraph, we first construct a KNN graph based on the euclidean distance in PCA space, and refine the edge weights between any two cells based on the shared overlap in their local neighborhoods (Jaccard similarity). This step is performed using the FindNeighbors function, and takes as input the previously defined dimensionality of the dataset (first 25 PCs).

To cluster the cells, we next apply modularity optimization techniques such as the Louvain algorithm (default) or SLM [SLM, Blondel et al., Journal of Statistical Mechanics], to iteratively group cells together, with the goal of optimizing the standard modularity function. The FindClusters function implements this procedure, and contains a resolution parameter that sets the 'granularity' of the downstream clustering, with increased values leading to a greater number of clusters. We find that setting this parameter between 0.4-1.2 typically returns good results for single-cell datasets of around 3K cells. Optimal resolution often increases for larger datasets. The clusters can be found using the Idents function.

```{r Graph based Clustering, eval=FALSE}
Sample_09_03_Tomato <- FindNeighbors(Sample_09_03_Tomato, dims = 1:30)
Sample_09_03_Tomato <- FindClusters(Sample_09_03_Tomato, resolution = 0.1)
Sample_09_03_Tomato <- FindClusters(Sample_09_03_Tomato, resolution = 0.2)
Sample_09_03_Tomato <- FindClusters(Sample_09_03_Tomato, resolution = 0.3)
```

## Run non-linear dimensional reduction (UMAP/tSNE)

Seurat offers several non-linear dimensional reduction techniques, such as tSNE and UMAP, to visualize and explore these datasets. The goal of these algorithms is to learn the underlying manifold of the data in order to place similar cells together in low-dimensional space. Cells within the graph-based clusters determined above should co-localize on these dimension reduction plots. As input to the UMAP and tSNE, we suggest using the same PCs as input to the clustering analysis.

```{r compute UMAP, eval=FALSE}
# If you haven't installed UMAP, you can do so via reticulate::py_install(packages =
# 'umap-learn')
Sample_09_03_Tomato <- RunUMAP(Sample_09_03_Tomato, dims = 1:30)
```


```{r plot UMAP}
# note that you can set `label = TRUE` or use the LabelClusters function to help label
# individual clusters
DimPlot(Sample_09_03_Tomato, reduction = "umap", group.by = "RNA_snn_res.0.3", label = T, label.size = 5)

```
```{r}
FeaturePlot(Sample_09_03_Tomato, 
            features = c("Rbpj",
                        "tdTomato-WPRE-sv40pA"))
VlnPlot(Sample_09_03_Tomato, 
            features = c("Rbpj",
                        "tdTomato-WPRE-sv40pA"),pt.size = 0)

```



```{r Gene and RNA counts}
# note that you can set `label = TRUE` or use the LabelClusters function to help label
# individual clusters
FeaturePlot(Sample_09_03_Tomato, reduction = "umap",
            features = c("nCount_RNA", "nFeature_RNA"))
```



# Find Doublets

I use here DoubletFinder for predcting doublets in the dataset

DoubletFinder can be broken up into 4 steps:
(1) Generate artificial doublets from existing scRNA-seq data
(2) Pre-process merged real-artificial data
(3) Perform PCA and use the PC distance matrix to find each cell's proportion of artificial k nearest neighbors (pANN)
(4) Rank order and threshold pANN values according to the expected number of doublets



```{r DoubletFinder 1, eval=FALSE}
## pK Identification (no ground-truth) ---------------------------------------------------------------------------------------
sweep.res.list_Tomato <- paramSweep_v3(Sample_09_03_Tomato, PCs = 1:10, sct = FALSE)
sweep.stats_Tomato <- summarizeSweep(sweep.res.list_Tomato, GT = FALSE)
bcmvn_Tomato <- find.pK(sweep.stats_Tomato)
```


```{r DoubletFinder 2, eval=FALSE}
## Homotypic Doublet Proportion Estimate -------------------------------------------------------------------------------------
homotypic.prop <- modelHomotypic(Sample_09_03_Tomato@meta.data$RNA_snn_res.0.2)           ## ex: annotations <- seu_kidney@meta.data$ClusteringResults
nExp_poi <- round(0.08*nrow(Sample_09_03_Tomato@meta.data))  ## Assuming 8% double-based on the cell loading
nExp_poi.adj <- round(nExp_poi*(1-homotypic.prop))
```



```{r DoubletFinder 3, eval=FALSE}
Sample_09_03_Tomato <- doubletFinder_v3(Sample_09_03_Tomato, PCs = 1:10,
                            pN = 0.25, pK = 0.09,
                            nExp = nExp_poi, reuse.pANN = FALSE,
                            sct = FALSE)
Sample_09_03_Tomato <- doubletFinder_v3(Sample_09_03_Tomato, PCs = 1:10, 
                            pN = 0.25, pK = 0.09,
                            nExp = nExp_poi.adj,
                            reuse.pANN = "pANN_0.25_0.09_913",
                            sct = FALSE)
```



```{r DoubletFinder 4}
p1 <- DimPlot(Sample_09_03_Tomato, group.by="DF.classifications_0.25_0.09_913",
        cols = c("red","grey"),
        reduction="umap", pt.size=0.3 
        )
p2 <- FeaturePlot(Sample_09_03_Tomato, reduction = "umap",
            features =  "nFeature_RNA")

p1 + p2
```


```{r doublet viz}
p1 <- DimPlot(Sample_09_03_Tomato, reduction = "umap", group.by = "RNA_snn_res.0.3", 
              label = T, label.size = 5) + theme(legend.position="none")
p2 <- FeaturePlot(Sample_09_03_Tomato, reduction = "umap", 
                  features = "pANN_0.25_0.09_913") + theme(legend.position="none")


p3 <- VlnPlot(Sample_09_03_Tomato, features =  "pANN_0.25_0.09_913",
              group.by =  "RNA_snn_res.0.3", pt.size = 0 ) + theme(legend.position="none")

p1 + p2 /p3
```

### Conclusions from doublet analysis
From this analysis it seems that (at this resolution) clusters 9, 19, 20 and 21 are clear doublets



# Check for some important markers

In this sample there is a mix of **Blood, Liver, Heart and Lung cells**

![Loaded cells](./loaded_cells.png){#id .class width=70%}

From this image we expect to have 
```{r cells loaded}

cells_loaded <- data.frame(tissue=c("Blood", "Liver","Heart", "Lung"),
                           count= c(2097,53000,23048,26461) )
cells_loaded$proportion <- cells_loaded$count/sum(cells_loaded$count)
ggplot(data=cells_loaded, aes(x=tissue, y=proportion, fill=tissue)) +
  geom_bar(stat="identity") +   theme_minimal()

```



## Blood cells

```{r Blood cells markers}
FeaturePlot(Sample_09_03_Tomato,
            features = c("Ptprc", # Leukocyte lineage
                         "Cd14", # Macrophage
                         "Cd3e", # T cell                        
                         "Cd79a") # B cell
            )
```
 
```{r}
DimPlot(Sample_09_03_Tomato, reduction = "umap", group.by = "RNA_snn_res.0.3", 
              label = T, label.size = 5) + theme(legend.position="none")
```

To mark all the cells that are of Leukocyte lineage, I use the
[Ptprc](https://www.uniprot.org/uniprot/P06800) gene, which expression is
restricted to the hematopoietic compartment of development.


```{r}
VlnPlot(Sample_09_03_Tomato, ncol = 2,
        features =  c("Ptprc",
                      "Cd14", # Macrophage
                      "Cd3e", # T cell                        
                      "Cd79a"), # B cell
        group.by =  "RNA_snn_res.0.3", 
        pt.size = 0 ) + theme(legend.position="none")
```

From this analysis it seems that clusters 1, 5, 6, 8, 12, 15, 17 are blood cells.

And from classic markers it seems that these are the specific cell clusters
T cells =    8
B cells =    1, 12, 15, 17
Macrophage = 5, 6 (19-21 have markers but are suspected doublets)


## Non-blood cells

Now I will plot only the clusters that were not positive for "Ptprc".


```{r}
x <- Sample_09_03_Tomato@meta.data$RNA_snn_res.0.3  %in%  as.character(c(0,2:4,7,9:11,13,14,16,18))
p1 <- DimPlot(Sample_09_03_Tomato, reduction = "umap", cells = x,
              group.by = "RNA_snn_res.0.3",
              label = T, pt.size = 0.5)
p2 <- DimPlot(Sample_09_03_Tomato,   cells = x,group.by="DF.classifications_0.25_0.09_913",
              cols = c("red","grey"),
              reduction="umap", pt.size=0.3 )
p1 + p2

```




## Find marker genes
```{r, eval=FALSE}
# find markers for every cluster compared to all remaining cells, report only the positive ones
Sample_09_03_Tomato.markers <- FindAllMarkers(Sample_09_03_Tomato,
                                           only.pos = TRUE, min.pct = 0.25,
                                           logfc.threshold = 0.25)
Sample_09_03_Tomato.markers %>% group_by(cluster) %>% top_n(n = 2, wt = avg_log2FC)
```



```{r Graph based Clustering nonblood}
x <- Sample_09_03_Tomato@meta.data$RNA_snn_res.0.3  %in%  as.character(c(0,2:4,7,9:11,13,14,16,18))
Sample_09_03_Tomato_nonblood <- subset(Sample_09_03_Tomato, cells = colnames(Sample_09_03_Tomato)[x])
Sample_09_03_Tomato_nonblood <- FindVariableFeatures(Sample_09_03_Tomato_nonblood, 
                                         selection.method = "vst",
                                         nfeatures = 2000)
Sample_09_03_Tomato_nonblood <- RunPCA(Sample_09_03_Tomato_nonblood,
                                  features = VariableFeatures(object = Sample_09_03_Tomato_nonblood))
Sample_09_03_Tomato_nonblood <- FindNeighbors(Sample_09_03_Tomato_nonblood, dims = 1:30)
Sample_09_03_Tomato_nonblood <- FindClusters(Sample_09_03_Tomato_nonblood, resolution = 0.1)
Sample_09_03_Tomato_nonblood <- FindClusters(Sample_09_03_Tomato_nonblood, resolution = 0.05)

Sample_09_03_Tomato_nonblood <- RunUMAP(Sample_09_03_Tomato_nonblood, dims = 1:30)

```


```{r plot UMAP nonblood}
# note that you can set `label = TRUE` or use the LabelClusters function to help label
# individual clusters
DimPlot(Sample_09_03_Tomato_nonblood, reduction = "umap", group.by = "RNA_snn_res.0.05", label = T, label.size = 5)

```



 
```{r Main cell types nonblood}
FeaturePlot(Sample_09_03_Tomato_nonblood,
            features = c("Cdh5", # endothelial marker 
                        "Bmp2", # Liver
                        "Emp2", # Lung 
                        "Col3a1") # heart
            )
```

```{r Main cells 2}
FeaturePlot(Sample_09_03_Tomato_nonblood,
            c("Clec1b", #liver
              "Fendrr", # lung
 #             "Ebf3",
              "Sparcl1", #Heart and muscle
              "Meox1") #Heart and muscle
                 ) 
```


### Some of the MARKER genes.
RSPO3 (Veins), Wnt2 (Venous capillaries), LTBP4 (Arterial capillaries), Gja5 (Arteries), Odc1 (Dll4 or Rbpj LOF Highly Metabolic cells), Mki67 (Proliferating Cells in G2/M) and MCM6 for cells in S-phase.


```{r Main cell types 2}
FeaturePlot(Sample_09_03_Tomato_nonblood, 
            features = c("Rspo3", # Veins
                        "Ltbp4", # Arterial capillaries 
                        "Mki67", # Proliferating Cells in G2/M
                        "Mcm6") #   cells in S-phase.
            
            )
```


# Open Kalucka dataset

The Kalucka dataset was downloaded from [the edothelial atlas](https://endotheliomics.shinyapps.io/ec_atlas/).
I downloaded the whole dataset, but I will subset the data to only include the organs we are interested on:
- Liver
- Heart
- Lung
```{r}
kalucka <- readRDS(file = "../rds/kalucka_norm.rds")
DimPlot(kalucka, group.by = "Tissue", label = T)

```


```{r}
kalucka_TOI <- kalucka@meta.data$Tissue  %in%  c("liver","lung","heart")

kalucka_subset <- subset(kalucka, cells = colnames(kalucka)[kalucka_TOI])
kalucka_subset <- FindNeighbors(kalucka_subset, dims = 1:10)
kalucka_subset <- FindClusters(kalucka_subset)

kalucka_subset <- RunUMAP(kalucka_subset, dims = 1:10, return.model = TRUE)
DimPlot(kalucka_subset, group.by ="Tissue", label = T)

```


# Mapping to Kalucka reference 

Once that I have subset the kalucka dataset, I will map Sample_09_03 to it.

First I find anchors between reference and query. 

```{r}
anchors <- FindTransferAnchors(
  reference = kalucka_subset,
  query = Sample_09_03_Tomato_nonblood,
  normalization.method = "LogNormalize",
  #reference.reduction = "pcaproject",
  dims = 1:50
)
```

We then transfer cell type labels from the reference to the query.
Additionally, we project the query data onto the UMAP structure of the reference.

```{r}
Sample_09_03_Tomato_nonblood <- TransferData(
  anchorset = anchors, 
  reference = kalucka_subset,
  query = Sample_09_03_Tomato_nonblood,
  refdata = list(
    tissue = "Tissue",
    #endoth = "Endothelial.cell",
    predicted_ADT = "ADT")
)
```



```{r}
Sample_09_03_Tomato_nonblood <- IntegrateEmbeddings(
  anchorset = anchors,
  reference = kalucka_subset,
  query = Sample_09_03_Tomato_nonblood, 
  new.reduction.name = "ref.spca"
)
```
```{r}
Sample_09_03_Tomato_nonblood <- ProjectUMAP(
  query = Sample_09_03_Tomato_nonblood, 
  query.reduction = "ref.spca",
  reference = kalucka_subset,
  reference.reduction = "pca", 
  reduction.model = "umap"
)
```
## Visualise integration results

Now we can see where the cells of the Sample_09_03 maps on to the Kalucka reference.
We can see that some clear clusters map to the reference, like cluster 0 to brain,
2 and 4 to lung.

```{r}
p1 = DimPlot(Sample_09_03_Tomato_nonblood, reduction = "umap",
             group.by = "predicted.tissue", pt.size = 0.5, 
             label = TRUE, label.size = 5, repel = TRUE)
p2 = DimPlot(Sample_09_03_Tomato_nonblood, reduction = "umap",
             group.by = "RNA_snn_res.0.05",  pt.size = 0.7, 
             label = TRUE, label.size = 5 ,repel = TRUE) + NoLegend()
p3 <- ggplot(data=cells_loaded, aes(x=tissue, y=proportion, fill=tissue)) +
      geom_bar(stat="identity") +   theme_minimal()
p4 <- dittoBarPlot(Sample_09_03_Tomato_nonblood, "predicted.tissue",
                   group.by = "RNA_snn_res.0.05")
p1 +  p2 
p3 + p4

```
```{r}
FeaturePlot(object = Sample_09_03_Tomato_nonblood, features = "Cdh5")
```
```{r}
FeaturePlot(Sample_09_03_Tomato_nonblood, features = "predicted.tissue.score")
```

Now we can see where the cells of the Sample_09_03_Tomato maps on to the Kalucka
reference.
We can see that some clear clusters map to the reference, like cluster 0 to liver,
3 to heart and 1 to lung.

According to the prediction, other clusters like 2 and 5 are heart, but these 
clusters are non-endothelial (Cdh5-) so the kalucka dataset is not the best, 
I would need to use organ specific datasets from the tabula muris or so.


# extract info on BC "genes"  
First we need to check if we got reads mapped to the Barcodes (cellranger count).
As the genes were appended to the fasta and gtf files, the Barcode genes should
appear at the end of the gene matrix.


```{r Show reads of BCs}
all.genes <- rownames(Sample_09_03_Tomato)
## Show the last 60 genes
tail(all.genes, n = 63) # ALL genes

```

```{r}
FeaturePlot(Sample_09_03_Tomato, 
            features = c("A1-BC01F",
                        "A1-BC10F",  
                        "A3-BC01F"))
VlnPlot(Sample_09_03_Tomato, 
            features = c("A1-BC01F",
                        "A1-BC10F",  
                        "A3-BC01F"))
```

```{r Count cells with BCs}
BC.matrix <- Sample_09_03_Tomato@assays$RNA@counts

# get BC genes 
BC_genes <- tail(all.genes, n = 13) # ALL genes
BC_genes <- BC_genes[1:12]

# subset the matrix to include only the shared genes
x = rownames(BC.matrix) %in% BC_genes
BC.matrix <- BC.matrix[x,]
dim(BC.matrix)
bc <- as.matrix(BC.matrix)
bc1f <- bc[, colSums(bc != 0 ) > 0 ]
bc2f <- bc[, colSums(bc != 0 ) > 1 ]

barplot(c(dim(bc1f)[2], dim(bc2f)[2]), names.arg = c("1 barcode", "2 Barcodes"),
        col = "#F8766D",
        legend.text = "Cells with at least 1 Barcode read" )

par(mar= c(8,4,4,2))
plot(apply(bc, 1, sum), main = "FRT Barcode reads", 
     ylab ="Reads count", axes=F,
     xlab = "")
axis(side = 1, labels = BC_genes, at = 1:12, las=3)
axis(side =2)
grid()
```


From this we can see that there are reads only from **3 BARCODES** and all from 1st array.

Now it is important to quantify the number of cells that have at least 1 read of
the Barcodes, and at least 2 of them.


# Save Object for further analyses

```{r Save Object for further analyses, eval=FALSE}

saveRDS(Sample_09_03_Tomato, "../rds/Sample_09_03_Tomato.rds")
saveRDS(Sample_09_03_Tomato_nonblood, "../rds/Sample_09_03_Tomato_nonblood.rds")

```

