---
title: "Sample_28Jul21"
author: "Irepan Salvador-Martinez"
date: "12/11/2021"
output:
  BiocStyle::html_document:
    toc: true
    toc_float: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
  echo = TRUE, out.width = "100%", fig.align = "center",
  message = FALSE, warning = FALSE
)
options(width = 900)
```


# 28Jul21 Samples ADULT MICE

This sample is a complex one.


## About input files

The files used as input for this analysis are the files coming out of the the
*cellranger multi* pipeline, which combines gene expression, antibody capture, 
and multiplexing barcodes (CMOs).

In this case I used cellranger v6.0.0.

To know more on [multi option see here](https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/using/multi).


**NOTE** Tho read the multi experiment see the [Seurat tutorial found here](https://satijalab.org/seurat/articles/multimodal_vignette.html#loading-data-from-10x-multi-modal-experiments-1)

The cellranger command is the usual one but in order to read our Recombinase
arrays and other transgenes (YFP, Tomato, etc) I needed to append their sequences into the fasta files and the IDs into
the gtf files (using the *cellranger mkref* pipeline).


```{r packages}
library(Matrix)
library(dplyr)
library(Seurat)
library(patchwork)
library(ggplot2)
library(DoubletFinder)
library(BiocStyle)
library(dittoSeq)
library(RColorBrewer)
```

```{r}

ggplotColours <- function(n = 6, h = c(0, 360) + 15){
  if ((diff(h) %% 360) < 1) h[2] <- h[2] - 360/n
  hcl(h = (seq(h[1], h[2], length = n)), c = 100, l = 65)
}

```


## Loading data from 10X multi-modal experiments

Seurat is also able to analyze data from multimodal 10X experiments processed using CellRanger v3.

I first read all the data and then create a seurat object with the gene expression, and append the antibody and multiplex assays

```{r Load precomputed data, eval=FALSE}
Sample_28Jul21 <-  readRDS("Sample_28Jul21.rds")
# define assay
DefaultAssay(Sample_28Jul21) <- "RNA"
```


```{r Read 10X data}
raw_dir <- "../Raw_Data/"

Sample_28Jul21.data <- Read10X(data.dir = raw_dir)

x <- CreateSeuratObject(counts = Sample_28Jul21.data[["Gene Expression"]], min.cells = 3, min.features = 200)
cells_OK <- colnames(x)
rm(x)

Sample_28Jul21 <- CreateSeuratObject(counts = Sample_28Jul21.data[["Gene Expression"]])
Sample_28Jul21[["ADT"]] <- CreateAssayObject(Sample_28Jul21.data[["Antibody Capture"]])
Sample_28Jul21[["CTO"]] <- CreateAssayObject(Sample_28Jul21.data[["Multiplexing Capture"]])

Sample_28Jul21 <- subset(Sample_28Jul21, cells = cells_OK)
dim(Sample_28Jul21)

savedir = "../rds/"
saveRDS(object = Sample_28Jul21, file = paste0(savedir,"Sample_28Jul21.rds") )
rm(Sample_28Jul21.data)
```


Show the list of features measured in the ADT assay

```{r}
rownames(Sample_28Jul21[["ADT"]])
```


# Standard pre-processing workflow

The steps below encompass the standard pre-processing workflow for scRNA-seq data in Seurat. These represent the selection and filtration of cells based on QC metrics, data normalization and scaling, and the detection of highly variable features.

### QC and selecting cells for further analysis

Seurat allows you to easily explore QC metrics and filter cells based on any user-defined criteria. A few QC metrics commonly used by the community include

  The number of unique genes detected in each cell.
    Low-quality cells or empty droplets will often have very few genes
    Cell doublets or multiplets may exhibit an aberrantly high gene count
  Similarly, the total number of molecules detected within a cell (correlates strongly with unique genes)
  The percentage of reads that map to the mitochondrial genome
    Low-quality / dying cells often exhibit extensive mitochondrial contamination
    We calculate mitochondrial QC metrics with the PercentageFeatureSet function, which calculates the percentage of counts originating   from a set of features
    We use the set of all genes starting with MT- as a set of mitochondrial genes

```{r Mt genes}
# The [[ operator can add columns to object metadata. This is a great place to stash QC stats
Sample_28Jul21[["percent.mt"]] <- PercentageFeatureSet(Sample_28Jul21, pattern = "^mt-")
# Show QC metrics for the first 5 cells
head(Sample_28Jul21@meta.data, 50)
```


In the example below, we visualize QC metrics, and use these to filter cells.

  We filter cells that have unique feature counts over 2,500 or less than 200
  We filter cells that have >10% mitochondrial counts



```{r QC metrics,fig.height=3}
# Visualize QC metrics as a violin plot
VlnPlot(Sample_28Jul21, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"),
        group.by = "orig.ident", ncol = 3, pt.size = 0)
```

```{r plot features,fig.height=4}
# FeatureScatter is typically used to visualize feature-feature relationships, but can be used
# for anything calculated by the object, i.e. columns in object metadata, PC scores etc.

plot1 <- FeatureScatter(Sample_28Jul21, feature1 = "nFeature_RNA",
                        group.by = "orig.ident", feature2 = "percent.mt")
plot2 <- FeatureScatter(Sample_28Jul21, feature1 = "nCount_RNA",
                        group.by = "orig.ident",feature2 = "nFeature_RNA")
plot1 / plot2
```

```{r,fig.height=4}
plot(x = Sample_28Jul21@meta.data$nCount_RNA,
    y = Sample_28Jul21@meta.data$nFeature_RNA, 
    type = "p",pch=20 , cex= 0.2,
    xlim = c(0,45000), ylim = c(200,6500), col="darkblue")+
  abline(h = c(500,6000), lty=2)+
  abline(v = c(1000,35000), lty=2)

```
```{r,fig.height=3}
plot(x = Sample_28Jul21@meta.data$nCount_RNA,
    y = Sample_28Jul21@meta.data$percent.mt,
    type = "p",pch=20 , cex= 0.2,xlim = c(0,45000), col="darkblue")+
abline(h = c(15), lty=2)+
abline(v = c(1000,35000), lty=2)
```

```{r Filter cells}
Sample_28Jul21 <- subset(Sample_28Jul21,
                       subset = nFeature_RNA > 500 & nFeature_RNA < 6000 &
                       nCount_RNA > 1000 & nCount_RNA < 35000  & percent.mt < 15)
```


## Normalizing the data

After removing unwanted cells from the dataset, the next step is to normalize the data. By default, we employ a global-scaling normalization method "LogNormalize" that normalizes the feature expression measurements for each cell by the total expression, multiplies this by a scale factor (10,000 by default), and log-transforms the result. Normalized values are stored in Sample_28Jul21[["RNA"]]@data.

```{r Normalise}
Sample_28Jul21 <- NormalizeData(Sample_28Jul21, assay = "RNA")
Sample_28Jul21 <- NormalizeData(Sample_28Jul21, assay = "ADT", normalization.method = "CLR")

```

## Identification of highly variable features (feature selection)

We next calculate a subset of features that exhibit high cell-to-cell variation in the dataset (i.e, they are highly expressed in some cells, and lowly expressed in others). We and others have found that focusing on these genes in downstream analysis helps to highlight biological signal in single-cell datasets.

Our procedure in Seurat3 is described in detail here, and improves on previous versions by directly modeling the mean-variance relationship inherent in single-cell data, and is implemented in the FindVariableFeatures function. By default, we return 2,000 features per dataset. These will be used in downstream analysis, like PCA.


```{r Highly variable genes}

Sample_28Jul21 <- FindVariableFeatures(Sample_28Jul21, selection.method = "vst", nfeatures = 3000)

```

```{r SHOW Highly variable,fig.height=3 }
# Identify the 10 most highly variable genes
top10 <- head(VariableFeatures(Sample_28Jul21), 10)

# plot variable features with and without labels
plot1 <- VariableFeaturePlot(Sample_28Jul21)
plot2 <- LabelPoints(plot = plot1, points = top10, repel = T, xnudge = 0, ynudge = 0)
plot2

```

## Scaling the data

Next, we apply a linear transformation ('scaling') that is a standard pre-processing step prior to dimensional reduction techniques like PCA. The ScaleData function:

  Shifts the expression of each gene, so that the mean expression across cells is 0
  Scales the expression of each gene, so that the variance across cells is 1
    This step gives equal weight in downstream analyses, so that highly-expressed genes do not dominate
  The results of this are stored in Sample_28Jul21[["RNA"]]@scale.data
```{r Scaling}
all.genes <- rownames(Sample_28Jul21)
Sample_28Jul21 <- ScaleData(Sample_28Jul21, features = all.genes)

```


## Perform linear dimensional reduction

Next we perform PCA on the scaled data. By default, only the previously determined variable features are used as input, but can be defined using features argument if you wish to choose a different subset.

```{r PCA}
Sample_28Jul21 <- RunPCA(Sample_28Jul21, features = VariableFeatures(object = Sample_28Jul21))
```

Seurat provides several useful ways of visualizing both cells and features that define the PCA, including VizDimReduction, DimPlot, and DimHeatmap

```{r}
# Examine and visualize PCA results a few different ways
print(Sample_28Jul21[["pca"]], dims = 1:5, nfeatures = 5)
```

```{r PCA genes,fig.height=4}
VizDimLoadings(Sample_28Jul21, dims = 1:2, reduction = "pca")
```


```{r PCA viz,fig.height=3}
DimPlot(Sample_28Jul21,group.by = "orig.ident", reduction = "pca")
```

```{r PCA 12 dims,fig.height=4}
DimHeatmap(Sample_28Jul21, dims = 1:6, cells = 500, balanced = TRUE)
```


## Determine the 'dimensionality' of the dataset

To overcome the extensive technical noise in any single feature for scRNA-seq data, Seurat clusters cells based on their PCA scores, with each PC essentially representing a 'metafeature' that combines information across a correlated feature set. The top principal components therefore represent a robust compression of the dataset. However, how many components should we choose to include? 10? 20? 100?


```{r Elbowplot,fig.height=3} 
ElbowPlot(Sample_28Jul21, ndims = 50)

```


## Cluster the cells

Seurat v3 applies a graph-based clustering approach, building upon initial strategies in (Macosko et al). Importantly, the distance metric which drives the clustering analysis (based on previously identified PCs) remains the same. However, the approach to partitioning the cellular distance matrix into clusters has dramatically improved. The approach was heavily inspired by recent manuscripts which applied graph-based clustering approaches to scRNA-seq data [SNN-Cliq, Xu and Su, Bioinformatics, 2015] and CyTOF data [PhenoGraph, Levine et al., Cell, 2015]. Briefly, these methods embed cells in a graph structure - for example a K-nearest neighbor (KNN) graph, with edges drawn between cells with similar feature expression patterns, and then attempt to partition this graph into highly interconnected 'quasi-cliques' or 'communities'.

As in PhenoGraph, we first construct a KNN graph based on the euclidean distance in PCA space, and refine the edge weights between any two cells based on the shared overlap in their local neighborhoods (Jaccard similarity). This step is performed using the FindNeighbors function, and takes as input the previously defined dimensionality of the dataset (first 25 PCs).

To cluster the cells, we next apply modularity optimization techniques such as the Louvain algorithm (default) or SLM [SLM, Blondel et al., Journal of Statistical Mechanics], to iteratively group cells together, with the goal of optimizing the standard modularity function. The FindClusters function implements this procedure, and contains a resolution parameter that sets the 'granularity' of the downstream clustering, with increased values leading to a greater number of clusters. We find that setting this parameter between 0.4-1.2 typically returns good results for single-cell datasets of around 3K cells. Optimal resolution often increases for larger datasets. The clusters can be found using the Idents function.

```{r Graph based Clustering}
Sample_28Jul21 <- FindNeighbors(Sample_28Jul21, dims = 1:30)
Sample_28Jul21 <- FindClusters(Sample_28Jul21, resolution = 0.1)
Sample_28Jul21 <- FindClusters(Sample_28Jul21, resolution = 0.2)
Sample_28Jul21 <- FindClusters(Sample_28Jul21, resolution = 0.3)
```
## Run non-linear dimensional reduction (UMAP/tSNE)

Seurat offers several non-linear dimensional reduction techniques, such as tSNE and UMAP, to visualize and explore these datasets. The goal of these algorithms is to learn the underlying manifold of the data in order to place similar cells together in low-dimensional space. Cells within the graph-based clusters determined above should co-localize on these dimension reduction plots. As input to the UMAP and tSNE, we suggest using the same PCs as input to the clustering analysis.

```{r compute UMAP}
# If you haven't installed UMAP, you can do so via reticulate::py_install(packages =
# 'umap-learn')
Sample_28Jul21 <- RunUMAP(Sample_28Jul21, dims = 1:30)
```


```{r Gene and RNA counts,fig.height=4}
# note that you can set `label = TRUE` or use the LabelClusters function to help label
# individual clusters
FeaturePlot(Sample_28Jul21, reduction = "umap",
            features = c("nCount_RNA", "nFeature_RNA"))
```


```{r plot UMAP}
# note that you can set `label = TRUE` or use the LabelClusters function to help label
# individual clusters
DimPlot(Sample_28Jul21, reduction = "umap", label = T, label.size = 5)

```


```{r,fig.height=8, fig.width=5}
# Now, we will visualize CD14 levels for RNA and protein By setting the default assay, we can
# visualize one or the other
DefaultAssay(Sample_28Jul21) <- "ADT"
p1 <- FeaturePlot(Sample_28Jul21, c("B0305","B0306","B0307"),
                  order = T,ncol = 1,
                  cols = c("lightgrey", "darkgreen")) 

p1
```
```{r,fig.height=8, fig.width=8}
# Now, we will visualize CD14 levels for RNA and protein By setting the default assay, we can
# visualize one or the other
DefaultAssay(Sample_28Jul21) <- "CTO"
p2 <- FeaturePlot(Sample_28Jul21, 
                  c("CMO305","CMO306","CMO307","CMO308","CMO309","CMO310","CMO311","CMO312"),ncol = 2,
                  order = T)

DefaultAssay(Sample_28Jul21) <- "RNA"
p2
```


```{r Count cells with BCs}
DefaultAssay(Sample_28Jul21) <- "RNA"
BC.matrix <- Sample_28Jul21@assays$RNA@counts
all.genes <- rownames(Sample_28Jul21)
# get BC genes 
BC_genes <- tail(all.genes, n = 27) # ALL genes
BC_genes <- BC_genes[1:24]
# subset the matrix to include only the shared genes
x = rownames(BC.matrix) %in% BC_genes
BC.matrix <- BC.matrix[x,]
dim(BC.matrix)
bc <- as.matrix(BC.matrix)
bc1f <- bc[, colSums(bc != 0 ) > 0 ]
bc2f <- bc[, colSums(bc != 0 ) > 1 ]
barplot(c(dim(bc1f)[2], dim(bc2f)[2]), names.arg = c("1 barcode", "2 barcodes"),
        col = "#F8766D",
        legend.text = "Cells with at least 1 Barcode read" )
```
```{r}
FeaturePlot(Sample_28Jul21, 
             c("A1-B1-B7-B8","A1-B1-B2-B8","A1-B1-B2-B6-B7-B8","A1-B1-B2-B3-B7-B8"),
             ncol = 2,
             order = T)
```

```{r Save precomputed data}
saveRDS(Sample_28Jul21, "../rds/Sample_28Jul21.Irepan.Retinas.rds")
```
